fs/zbd_zenfs.cc

```diff
@@ -492,193 +469,306 @@ IOStatus ZonedBlockDevice::AllocateMetaZone(Zone **out_meta_zone) {
+          : ZENFS_LABEL_DETAILED(IO_ALLOC, NON_WAL, LATENCY),
+      Env::Default());
+  metrics_->ReportQPS(ZENFS_LABEL(IO_ALLOC, QPS), 1);
+
+  // Check if a deferred IO error was set
+  s = GetZoneDeferredStatus();
+  if (!s.ok()) {
+    return s;
+  }
+
+  if (io_type != IOType::kWAL) {
+    s = ApplyFinishThreshold();
+    if (!s.ok()) {
+      return s;
+    }
+  }
+
+  WaitForOpenIOZoneToken(io_type == IOType::kWAL);
+
+  /* Try to fill an already open zone(with the best life time diff) */
+  s = GetBestOpenZoneMatch(file_lifetime, &best_diff, &allocated_zone);
+  if (!s.ok()) {
+    PutOpenIOZoneToken();
+    return s;
+  }
+
+  // Holding allocated_zone if != nullptr
+
+  if (best_diff >= LIFETIME_DIFF_COULD_BE_WORSE) {
+    if (io_type == IOType::kWAL && allocated_zone != nullptr) {
+      fprintf(stdout,
+              "\n Did not find a good zone for WAL allocation. Best diff: %d\n",
+              (int)best_diff);
+    }
+
+    bool got_token = GetActiveIOZoneTokenIfAvailable();
+
+    /* If we did not get a token, try to use the best match, even if the life
+     * time diff not good but a better choice than to finish an existing zone
+     * and open a new one
+     */
+    if (allocated_zone != nullptr) {
+      if (!got_token && best_diff == LIFETIME_DIFF_COULD_BE_WORSE) {
+        Debug(logger_,
+              "Allocator: avoided a finish by relaxing lifetime diff "
+              "requirement\n");
+      } else {
+        s = allocated_zone->CheckRelease();
+        if (!s.ok()) {
+          PutOpenIOZoneToken();
+          return s;
         }
+        allocated_zone = nullptr;
       }
     }
-  }
 
-  if (finish_victim != nullptr) {
-    IOStatus status = finish_victim->CheckRelease();
-    if (!status.ok()) return status;
-    finish_victim = nullptr;
+    /* If we haven't found an open zone to fill, open a new zone */
+    if (allocated_zone == nullptr) {
+      /* We have to make sure we can open an empty zone */
+      while (!got_token && !GetActiveIOZoneTokenIfAvailable()) {
+        s = FinishCheapestIOZone();
+        if (!s.ok()) {
+          PutOpenIOZoneToken();
+          return s;
+        }
+      }
+
+      s = AllocateEmptyZone(&allocated_zone);
+      if (!s.ok()) {
+        PutActiveIOZoneToken();
+        PutOpenIOZoneToken();
+        return s;
+      }
+
+      if (allocated_zone != nullptr) {
+        assert(allocated_zone->IsBusy());
+        allocated_zone->lifetime_ = file_lifetime;
+        new_zone = true;
+      }
+    }
   }
 
   if (allocated_zone) {
-    ok = allocated_zone->IsBusy();
-    assert(ok);
-    open_io_zones_++;
+    assert(allocated_zone->IsBusy());
     Debug(logger_,
           "Allocating zone(new=%d) start: 0x%lx wp: 0x%lx lt: %d file lt: %d\n",
           new_zone, allocated_zone->start_, allocated_zone->wp_,
           allocated_zone->lifetime_, file_lifetime);
+  } else {
+    PutOpenIOZoneToken();
   }
 
-  io_zones_mtx.unlock();
   LogZoneStats();
```

Comment:
Shall we only log zone stats in Non-WAL allocation threads?

Reply:
yes, that is a good idea